<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>2. Description of Inputs &mdash; coralME 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="3. Arquitecture of coralME" href="coralMEArquitecture.html" />
    <link rel="prev" title="1. Getting started" href="GettingStarted.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            coralME
          </a>
              <div class="version">
                1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="GettingStarted.html">1. Getting started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">2. Description of Inputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Types-of-inputs">2.1. Types of inputs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Required">2.1.1. Required</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Optional">2.1.2. Optional</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Configuration">2.1.3. Configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Description">2.2. Description</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Genome-(genome.gb)">2.2.1. Genome (genome.gb)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Requirements">2.2.1.1. Requirements</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#M-model-(m_model.json)">2.2.2. M-model (m_model.json)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id1">2.2.2.1. Requirements</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Gene-dictionary-(genes.txt)-[optional]">2.2.3. Gene dictionary (genes.txt) [optional]</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id2">2.2.3.1. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Notes">2.2.3.2. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Proteins-(proteins.txt)-[optional]">2.2.4. Proteins (proteins.txt) [optional]</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id3">2.2.4.1. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id4">2.2.4.2. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#RNAs-(RNAs.txt)-[optional]">2.2.5. RNAs (RNAs.txt) [optional]</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id5">2.2.5.1. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id6">2.2.5.2. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#TUs-(TUs.txt)-[optional]">2.2.6. TUs (TUs.txt) [optional]</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id7">2.2.6.1. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id8">2.2.6.2. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Gene-sequences-(sequences.fasta)-[optional]">2.2.7. Gene sequences (sequences.fasta) [optional]</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id9">2.2.7.1. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id10">2.2.7.2. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Configuration-of-paths-to-files-(inputs.json)">2.2.8. Configuration of paths to files (inputs.json)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id11">2.2.8.1. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id12">2.2.8.2. Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Configuration-of-parameters-(organism.json)">2.2.9. Configuration of parameters (organism.json)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id13">2.2.9.1. Requirements</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id14">2.2.9.2. Notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="coralMEArquitecture.html">3. Arquitecture of coralME</a></li>
<li class="toctree-l1"><a class="reference internal" href="ManualCuration.html">4. How to manually curate a ME-model using coralME</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">5. Frequently Asked Questions</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">coralME</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><span class="section-number">2. </span>Description of Inputs</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/BasicInputs.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Description-of-Inputs">
<h1><span class="section-number">2. </span>Description of Inputs<a class="headerlink" href="#Description-of-Inputs" title="Permalink to this heading"></a></h1>
<p>coralME takes a total of 7 inputs, 2 required and 5 optional:</p>
<section id="Types-of-inputs">
<h2><span class="section-number">2.1. </span>Types of inputs<a class="headerlink" href="#Types-of-inputs" title="Permalink to this heading"></a></h2>
<section id="Required">
<h3><span class="section-number">2.1.1. </span>Required<a class="headerlink" href="#Required" title="Permalink to this heading"></a></h3>
<ol class="arabic simple">
<li><p><strong>Genome file</strong> (genome.gb)</p></li>
<li><p><strong>M-model</strong> (m_model.json or m_model.xml)</p></li>
</ol>
</section>
<section id="Optional">
<h3><span class="section-number">2.1.2. </span>Optional<a class="headerlink" href="#Optional" title="Permalink to this heading"></a></h3>
<p>Downloadable from an existing <strong>BioCyc</strong> database under Special SmartTables. If no optional files are provided, coralME complements them with genome.gb</p>
<ol class="arabic simple" start="3">
<li><p><strong>Genes file</strong>, by default:genes.txt</p></li>
<li><p><strong>RNAs file</strong>, by default:RNAs.txt</p></li>
<li><p><strong>Proteins file</strong>, by default:proteins.txt</p></li>
<li><p><strong>TUs file</strong>, by default:TUs.txt.</p></li>
<li><p><strong>Sequences file</strong>, by default:sequences.fasta</p></li>
</ol>
</section>
<section id="Configuration">
<h3><span class="section-number">2.1.3. </span>Configuration<a class="headerlink" href="#Configuration" title="Permalink to this heading"></a></h3>
<ol class="arabic simple" start="8">
<li><p><strong>inputs.json</strong></p></li>
<li><p><strong>orgnanism.json</strong></p></li>
</ol>
<p><img alt="Drawing" src="_images/inputs.png" /></p>
</section>
</section>
<section id="Description">
<h2><span class="section-number">2.2. </span>Description<a class="headerlink" href="#Description" title="Permalink to this heading"></a></h2>
<section id="Genome-(genome.gb)">
<h3><span class="section-number">2.2.1. </span>Genome (genome.gb)<a class="headerlink" href="#Genome-(genome.gb)" title="Permalink to this heading"></a></h3>
<p>The genome file contains provides coralME with: * Gene annotations. * Gene sequences.</p>
<section id="Requirements">
<h4><span class="section-number">2.2.1.1. </span>Requirements<a class="headerlink" href="#Requirements" title="Permalink to this heading"></a></h4>
<ol class="arabic simple">
<li><p><strong>Locus tags (locus_tag or old_locus_tag) MUST be consistent with m_model.json</strong>. Make sure you download the same genome file that was used to reconstruct the M-model.</p></li>
<li><p>Has name genome.gb.</p></li>
<li><p>Genbank-compliant file. Must be read by BioPython correctly.</p></li>
<li><p>It must contain the entire genome sequence. Make sure to enable Customize View&gt;Show Sequence before downloading the genbank file from NCBI.</p></li>
</ol>
<p>See an example of <a class="reference external" href="./helper_files/pputida_tutorial/inputs/genome.gb">genome.gb</a> and <a class="reference external" href="./helper_files/pputida_tutorial/inputs/sequences.fasta">sequences.fasta</a></p>
</section>
</section>
<section id="M-model-(m_model.json)">
<h3><span class="section-number">2.2.2. </span>M-model (m_model.json)<a class="headerlink" href="#M-model-(m_model.json)" title="Permalink to this heading"></a></h3>
<p>The M-model provides coralME with the metabolic model components: * Metabolic network (M-matrix) * Gene-protein-reaction associations * Environmental and internal constraints * Reaction subsystems * Biomass composition</p>
<section id="id1">
<h4><span class="section-number">2.2.2.1. </span>Requirements<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h4>
<p>This file should meet the following requirements: 1. <strong>Gene identifiers MUST be consistent with genome.gb locus_tag or old_locus_tag</strong>. Make sure you download the same genome file that was used to reconstruct the M-model. 2. Has name m_model.json. 3. COBRApy-compliant. Must be read by cobrapy-0.25.0.</p>
<p>See an example of <a class="reference external" href="./helper_files/pputida_tutorial/inputs/m_model.json">m_model.json</a></p>
</section>
</section>
<section id="Gene-dictionary-(genes.txt)-[optional]">
<h3><span class="section-number">2.2.3. </span>Gene dictionary (genes.txt) [optional]<a class="headerlink" href="#Gene-dictionary-(genes.txt)-[optional]" title="Permalink to this heading"></a></h3>
<p>genes.txt is a gene information table that can be downloaded from the All genes of organism SmartTable of the <a class="reference external" href="https://biocyc.org/">BioCyc</a> database. Click Export&gt;to Spreadsheet File&gt;frame IDs. This file is optional and is meant to complement the information from genome.gb in case the latter is missing genes.</p>
<p>genes.txt provides coralME with: * Gene locus tags * Gene names * Gene annotations * Gene positions * Gene products (protein, tRNA, etc.)</p>
<section id="id2">
<h4><span class="section-number">2.2.3.1. </span>Requirements<a class="headerlink" href="#id2" title="Permalink to this heading"></a></h4>
<p>This file should meet the following requirements: 1. Contains the index Gene Name and columns Accession-1, Left-End-Position, Right-End-Position, and Product. 2. <strong>Accession-1 MUST be consistent with the gene IDs in the GPRs of m_model.json and with the locus_tag (or old_locus_tag) in genome.gb</strong>. 3. Gene Name is consistent with: * Column Genes of polypeptide, complex, or RNA of proteins.txt * Column Gene of RNAs.txt * Column Genes of transcription unit of TUs.txt * Gene identifiers in
sequences.fasta 4. Product is consistent with: * Index of proteins.txt * Index of RNAs.txt 5. Must be tab-separated</p>
</section>
<section id="Notes">
<h4><span class="section-number">2.2.3.2. </span>Notes<a class="headerlink" href="#Notes" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p>Requirements 3, 4 and 5 regarding ID consistency should be directly met if the files are downloaded from the same BioCyc database.</p></li>
<li><p>Left-End-Position and Right-End-Position do not need to be consistent with the positions genome.gb. coralME will keep the positions in genome.gb over the specified ones in genes.txt</p></li>
</ul>
<p>See an example of <a class="reference external" href="./helper_files/pputida_tutorial/inputs/genes.txt">genes.txt</a></p>
</section>
</section>
<section id="Proteins-(proteins.txt)-[optional]">
<h3><span class="section-number">2.2.4. </span>Proteins (proteins.txt) [optional]<a class="headerlink" href="#Proteins-(proteins.txt)-[optional]" title="Permalink to this heading"></a></h3>
<p>proteins.txt is a protein complex information table that can be downloaded from the All proteins of organism SmartTable of the <a class="reference external" href="https://biocyc.org/">BioCyc</a> database. Click Export&gt;to Spreadsheet File&gt;frame IDs. This file is optional and is meant to complement the information from genome.gb.</p>
<p>proteins.txt provides coralME with: * Protein complex compositions</p>
<section id="id3">
<h4><span class="section-number">2.2.4.1. </span>Requirements<a class="headerlink" href="#id3" title="Permalink to this heading"></a></h4>
<p>This file should meet the following requirements: 1. Contains the index (Proteins Complexes) and columns Common-Name, Genes of polypeptide, complex, or RNA, and Locations. 2. (Proteins Complexes) is consistent with: * Column Product of genes.txt 3. Genes of polypeptide, complex, or RNA is consistent with: * Index Gene Name of genes.txt 4. Must be tab-separated</p>
</section>
<section id="id4">
<h4><span class="section-number">2.2.4.2. </span>Notes<a class="headerlink" href="#id4" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p>Requirements 2, 3 and 4 regarding ID consistency should be directly met if the files are downloaded from the same BioCyc database.</p></li>
</ul>
<p>See an example of <a class="reference external" href="./helper_files/pputida_tutorial/inputs/proteins.txt">proteins.txt</a></p>
</section>
</section>
<section id="RNAs-(RNAs.txt)-[optional]">
<h3><span class="section-number">2.2.5. </span>RNAs (RNAs.txt) [optional]<a class="headerlink" href="#RNAs-(RNAs.txt)-[optional]" title="Permalink to this heading"></a></h3>
<p>RNAs.txt is an RNA annotation table that can be downloaded from the All RNAs of organism SmartTable of the <a class="reference external" href="https://biocyc.org/">BioCyc</a> database. Click Export&gt;to Spreadsheet File&gt;frame IDs. This file is optional and is meant to complement the information from genome.gb.</p>
<p>RNAs.txt provides coralME with: * Genes annotated as RNA products (e.g. tRNA, rRNA, etc.) * RNA gene annotations (e.g. amino acids - tRNA associations)</p>
<section id="id5">
<h4><span class="section-number">2.2.5.1. </span>Requirements<a class="headerlink" href="#id5" title="Permalink to this heading"></a></h4>
<p>This file should meet the following requirements: 1. Contains the index (All-tRNAs Misc-RNAs rRNAs) and columns Common-Name, and Gene 2. (All-tRNAs Misc-RNAs rRNAs) is consistent with: * Column Product of genes.txt 3. Gene is consistent with: * Index Gene Name of genes.txt 4. Must be tab-separated</p>
</section>
<section id="id6">
<h4><span class="section-number">2.2.5.2. </span>Notes<a class="headerlink" href="#id6" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p>Requirements 2, 3 and 4 regarding ID consistency should be directly met if the files are downloaded from the same BioCyc database.</p></li>
</ul>
<p>See an example of <a class="reference external" href="./helper_files/pputida_tutorial/inputs/RNAs.txt">RNAs.txt</a></p>
</section>
</section>
<section id="TUs-(TUs.txt)-[optional]">
<h3><span class="section-number">2.2.6. </span>TUs (TUs.txt) [optional]<a class="headerlink" href="#TUs-(TUs.txt)-[optional]" title="Permalink to this heading"></a></h3>
<p>TUs.txt is a transcription unit annotation table that can be downloaded from the All TUs of organism SmartTable of the <a class="reference external" href="https://biocyc.org/">BioCyc</a> database. Click Export&gt;to Spreadsheet File&gt;frame IDs. This file is optional and is meant to complement the information from genome.gb.</p>
<p>TUs.txt provides coralME with: * Co-transcribed genes (operons). * Direction of transcription. * TU IDs.</p>
<section id="id7">
<h4><span class="section-number">2.2.6.1. </span>Requirements<a class="headerlink" href="#id7" title="Permalink to this heading"></a></h4>
<p>This file should meet the following requirements: 1. Contains the index Transcription-Units and columns Genes of transcription unit, and Direction 2. Genes of transcription unit is consistent with: * Index Gene Name of genes.txt 3. Must be tab-separated</p>
</section>
<section id="id8">
<h4><span class="section-number">2.2.6.2. </span>Notes<a class="headerlink" href="#id8" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p>Requirements 2 and 3 regarding ID consistency should be directly met if the files are downloaded from the same BioCyc database.</p></li>
</ul>
<p>See an example of <a class="reference external" href="./helper_files/pputida_tutorial/inputs/TUs.txt">TUs.txt</a></p>
</section>
</section>
<section id="Gene-sequences-(sequences.fasta)-[optional]">
<h3><span class="section-number">2.2.7. </span>Gene sequences (sequences.fasta) [optional]<a class="headerlink" href="#Gene-sequences-(sequences.fasta)-[optional]" title="Permalink to this heading"></a></h3>
<p>sequences.fasta is a nucleotide FASTA file that can be downloaded from the All genes of organism SmartTable of the <a class="reference external" href="https://biocyc.org/">BioCyc</a> database. Click Export&gt;FASTA&gt;Find sequences. This file is optional and is meant to complement the information from genome.gb in case the latter is missing genes.</p>
<p>sequences.fasta provides coralME with: * Gene sequences</p>
<section id="id9">
<h4><span class="section-number">2.2.7.1. </span>Requirements<a class="headerlink" href="#id9" title="Permalink to this heading"></a></h4>
<div class="line-block">
<div class="line">This file should meet the following requirements: 1. Gene identifiers are consistent with: * Index Gene Name of genes.txt</div>
<div class="line">2. Must be tab-separated</div>
</div>
</section>
<section id="id10">
<h4><span class="section-number">2.2.7.2. </span>Notes<a class="headerlink" href="#id10" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p>Requirements 1, 2 and 3 regarding ID consistency should be directly met if the files are downloaded from the same BioCyc database.</p></li>
</ul>
<p>See an example of <a class="reference external" href="./helper_files/pputida_tutorial/inputs/sequences.fasta">sequences.fasta</a></p>
</section>
</section>
<section id="Configuration-of-paths-to-files-(inputs.json)">
<h3><span class="section-number">2.2.8. </span>Configuration of paths to files (inputs.json)<a class="headerlink" href="#Configuration-of-paths-to-files-(inputs.json)" title="Permalink to this heading"></a></h3>
<p>inputs.json is a JSON file containing paths to input files for coralME.</p>
<p>inputs.json provides coralME with: * Paths to input files</p>
<section id="id11">
<h4><span class="section-number">2.2.8.1. </span>Requirements<a class="headerlink" href="#id11" title="Permalink to this heading"></a></h4>
<p>This file should meet the following requirements: 1. Must be JSON-compliant 2. Must contain paths to required files (M-model and Genome). 3. All defined files must exist.</p>
</section>
<section id="id12">
<h4><span class="section-number">2.2.8.2. </span>Notes<a class="headerlink" href="#id12" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p>See an example of <a class="reference external" href="./helper_files/pputida_tutorial/input.json">input.json</a></p></li>
</ul>
</section>
</section>
<section id="Configuration-of-parameters-(organism.json)">
<h3><span class="section-number">2.2.9. </span>Configuration of parameters (organism.json)<a class="headerlink" href="#Configuration-of-parameters-(organism.json)" title="Permalink to this heading"></a></h3>
<p>organism.json is a JSON file containing paths to input files for coralME.</p>
<p>organism.json provides coralME with: * ME-modeling parameters</p>
<section id="id13">
<h4><span class="section-number">2.2.9.1. </span>Requirements<a class="headerlink" href="#id13" title="Permalink to this heading"></a></h4>
<p>This file should meet the following requirements: 1. Must be JSON-compliant 2. Must contain the standard fields.</p>
</section>
<section id="id14">
<h4><span class="section-number">2.2.9.2. </span>Notes<a class="headerlink" href="#id14" title="Permalink to this heading"></a></h4>
<ul class="simple">
<li><p>See an example of <a class="reference external" href="./helper_files/pputida_tutorial/organism.json">organism.json</a></p></li>
</ul>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="GettingStarted.html" class="btn btn-neutral float-left" title="1. Getting started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="coralMEArquitecture.html" class="btn btn-neutral float-right" title="3. Arquitecture of coralME" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Juan D. Tibocha-Bonilla, Rodrigo Santibanez-Palominos.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>