Search.setIndex({"docnames": ["BasicInputs", "FAQ", "GettingStarted", "ManualCuration", "coralMEArquitecture", "index"], "filenames": ["BasicInputs.ipynb", "FAQ.ipynb", "GettingStarted.ipynb", "ManualCuration.ipynb", "coralMEArquitecture.ipynb", "index.rst"], "titles": ["<span class=\"section-number\">2. </span>Description of Inputs", "<span class=\"section-number\">5. </span>Frequently Asked Questions", "<span class=\"section-number\">1. </span>Getting started", "<span class=\"section-number\">4. </span>How to manually curate a ME-model using coralME", "<span class=\"section-number\">3. </span>Arquitecture of coralME", "Welcome to coralME documentation!"], "terms": {"coralm": 0, "take": 0, "total": 0, "7": 0, "2": 0, "5": [0, 2], "xml": 0, "download": [0, 1, 5], "from": [0, 1, 3, 4, 5], "an": [0, 1, 2, 4, 5], "exist": [0, 5], "biocyc": [0, 5], "databas": [0, 1, 4], "under": 0, "special": [0, 3], "smarttabl": [0, 1], "If": [0, 1, 3, 4], "ar": [0, 2, 3, 4, 5], "provid": [0, 4], "complement": [0, 4, 5], "them": [0, 2, 3], "default": [0, 2], "orgnan": 0, "The": [0, 2, 4, 5], "contain": [0, 1, 3, 4], "annot": 0, "locu": 0, "tag": 0, "locus_tag": [0, 1], "old_locus_tag": 0, "must": [0, 1, 2], "consist": [0, 2, 5], "make": [0, 1, 3, 4], "sure": [0, 1], "you": [0, 1, 2, 3], "same": [0, 1, 2, 4], "wa": [0, 1, 4], "us": [0, 1, 2, 4, 5], "reconstruct": [0, 1, 4, 5], "ha": [0, 4, 5], "name": [0, 2, 4], "genbank": [0, 2], "compliant": 0, "read": [0, 4], "biopython": 0, "correctli": 0, "It": [0, 1, 3, 4], "entir": [0, 4], "enabl": 0, "custom": 0, "view": 0, "show": [0, 2, 3], "befor": [0, 3], "ncbi": 0, "see": [0, 2], "exampl": [0, 2], "metabol": 0, "compon": [0, 3], "network": 0, "matrix": [0, 3, 5], "reaction": [0, 3], "associ": [0, 3], "environment": 0, "intern": 0, "constraint": 0, "subsystem": [0, 3], "biomass": 0, "composit": [0, 3], "thi": [0, 1, 2, 4, 5], "should": [0, 2, 5], "meet": 0, "follow": [0, 1, 3], "1": 0, "identifi": [0, 4, 5], "3": 0, "cobrapi": 0, "0": [0, 2], "25": 0, "i": [0, 2, 4, 5], "inform": [0, 1, 2, 3, 4, 5], "tabl": 0, "can": [0, 1, 2, 3, 4], "all": [0, 1, 2, 3, 4], "click": 0, "export": 0, "spreadsheet": 0, "frame": 0, "id": [0, 1, 2, 3], "meant": [0, 3], "case": [0, 1], "latter": 0, "miss": 0, "posit": [0, 4], "product": 0, "trna": [0, 2, 3], "etc": 0, "index": [0, 2, 5], "column": [0, 1, 2, 3], "access": [0, 1, 4], "left": 0, "end": 0, "right": [0, 2], "gpr": 0, "polypeptid": [0, 2], "complex": [0, 2, 3], "transcript": [0, 2, 3], "unit": [0, 2, 3], "4": [0, 4], "tab": 0, "separ": 0, "regard": 0, "directli": [0, 3], "met": 0, "do": [0, 2, 5], "need": [0, 2, 3], "keep": [0, 3], "over": 0, "specifi": 0, "ones": [0, 3], "common": [0, 2], "locat": [0, 2, 3], "e": [0, 1, 3, 4, 5], "g": [0, 1, 4], "rrna": [0, 2, 3], "amino": [0, 3], "acid": [0, 3], "misc": [0, 2], "co": 0, "transcrib": 0, "operon": 0, "direct": [0, 2], "nucleotid": 0, "find": 0, "defin": [0, 3], "me": [0, 4, 5], "standard": [0, 2], "field": 0, "we": [1, 2, 4], "know": 1, "convent": 1, "problem": 1, "across": 1, "platform": 1, "bioinformat": 1, "tri": 1, "gener": [1, 3, 4, 5], "much": 1, "possibl": 1, "could": [1, 2], "often": 1, "differ": 1, "genom": 1, "assembli": 1, "yield": 1, "inconsist": 1, "depend": 1, "your": [1, 3, 5], "so": 1, "classifi": [1, 3], "issu": 1, "consid": [1, 4], "three": 1, "main": [1, 2, 4, 5], "sourc": 1, "m_model": 1, "json": [1, 3], "gb": 1, "": [1, 2, 5], "info": 1, "txt": [1, 3], "overal": 1, "assum": 1, "modifi": [1, 2, 3], "hardest": 1, "approach": 1, "thu": 1, "last": 1, "resort": 1, "look": [1, 3], "correct": [1, 3], "which": [1, 3], "correspond": 1, "One": 1, "quick": 1, "wai": [1, 2], "ensur": [1, 2], "copi": [1, 2, 3], "one": [1, 2, 3], "past": 1, "search": [1, 2, 5], "bar": [1, 2], "best": [1, 4], "scenario": 1, "microb": 1, "appear": [1, 2], "list": [1, 2], "solv": 1, "That": 1, "didn": 1, "t": 1, "help": 1, "even": 1, "when": [1, 3], "still": 1, "howev": 1, "somewher": 1, "sinc": 1, "found": 1, "try": 1, "ad": [1, 2, 3], "new": [1, 3], "synonym": 1, "mayb": 1, "onli": 1, "underscor": 1, "pp0001": 1, "pp_0001": 1, "text": 1, "editor": 1, "chang": [1, 3], "accordingli": [1, 3], "mistak": 1, "edit": [1, 2], "critic": 1, "happen": 1, "probabl": 1, "have": [1, 2, 3], "wrong": 1, "dictionari": 1, "convert": 1, "between": [1, 3, 4], "For": 2, "detail": 2, "descript": 2, "about": [2, 4, 5], "architectur": [2, 5], "arquitectur": [2, 5], "option": 2, "after": [2, 3], "m": [2, 3], "quickest": 2, "avail": [2, 4, 5], "To": 2, "here": [2, 3, 4], "process": [2, 3, 4, 5], "appli": 2, "some": 2, "dropdown": 2, "add": [2, 3], "properti": 2, "transform": 2, "within": 2, "webpag": 2, "note": [2, 3, 5], "awai": 2, "But": 2, "onc": 2, "stage": [2, 3], "applic": [2, 4], "mebuild": [2, 5], "org": [2, 4], "helper_fil": 2, "pputida_tutori": 2, "load": 2, "configur": 2, "o": 2, "chdir": 2, "format": [2, 3], "autoclass": 2, "noindex": 2, "generate_fil": 2, "overwrit": [2, 3], "true": [2, 4], "build_me_model": [2, 4], "fals": 2, "growth_key_and_valu": 2, "me_model": 2, "mu": 2, "001": 2, "set": [2, 3, 4], "valu": [2, 4], "feasibl": 2, "check": 2, "feel": 2, "free": 2, "sometim": 2, "too": 2, "high": 2, "put": 2, "signific": 2, "strain": 2, "give": 2, "mani": [2, 3], "gap": [2, 3], "low": 2, "might": [2, 3], "how": [2, 5], "run": 3, "first": 3, "time": 3, "building_data": 3, "most": 3, "also": 3, "automat": [3, 5], "fill": 3, "algorithm": [3, 5], "save": 3, "reference_fil": 3, "These": 3, "refer": [3, 4], "guid": [3, 5], "thei": 3, "map": 3, "doe": 3, "ani": [3, 4, 5], "alwai": 3, "termination_subreact": 3, "translat": 3, "termin": 3, "subreact": 3, "machineri": 3, "peptide_release_factor": 3, "peptid": 3, "releas": 3, "factor": 3, "rna_degradosom": 3, "rna": 3, "degradosom": 3, "special_trna_subreact": 3, "sec": 3, "selenocystein": 3, "synthesi": 3, "ser": 3, "lipoprotein_precursor": 3, "lipoprotein": 3, "precursor": 3, "special_modif": 3, "modif": 3, "pre": 3, "excision_machineri": 3, "excis": 3, "orphan_and_spont_react": 3, "mark": 3, "orphan": 3, "spontan": 3, "cplx_dummi": 3, "requir": [3, 5], "enzym": [3, 4], "flux": 3, "enzyme_reaction_associ": 3, "creat": 3, "peptide_compartment_and_pathwai": 3, "protein": 3, "transloc": 3, "pathwai": 3, "translocation_pathwai": 3, "rna_modif": 3, "perform": [3, 5], "either": 3, "ribosomal_protein": 3, "ribosom": 3, "rho_independ": 3, "gene": [3, 5], "rho": 3, "independ": 3, "sigma_factor": 3, "n": 3, "methionin": 3, "cleavag": 3, "cleaved_methionin": 3, "folding_dict": 3, "fold": 3, "translocation_multipli": 3, "pore": 3, "subreaction_matrix": 3, "me_metabolit": 3, "metabolit": 3, "replac": 3, "elongation_subreact": 3, "elong": 3, "subsystem_classif": 3, "umbrella": 3, "classif": 3, "median": 3, "keff": 3, "sasa": 3, "reaction_matrix": 3, "definit": 3, "lipid_modif": 3, "lipid": 3, "amino_acid_trna_synthetas": 3, "ligas": 3, "initiation_subreact": 3, "initi": [3, 5], "post_transcriptional_modification_of_rna": 3, "undergo": 3, "protein_correct": 3, "well": 3, "remov": [3, 5], "reaction_median_keff": 3, "transcription_subreact": 3, "generic_dict": 3, "ribosome_subreact": 3, "reaction_correct": 3, "build": [3, 5], "tus_from_biocyc": 3, "yet": 3, "go": 3, "back": 3, "gettingstart": 3, "ipynb": 3, "through": 3, "import": 3, "flag": 3, "risen": 3, "curation_not": 3, "further": 3, "everytim": 3, "troubleshoot": 3, "remain": 3, "warn": 3, "iter": 3, "finish": 3, "present": 3, "irrelev": 3, "compos": 4, "class": 4, "exchang": 4, "specif": [4, 5], "model": [4, 5], "role": 4, "store": 4, "act": 4, "necessari": [4, 5], "retriev": 4, "ref": 4, "manipul": 4, "method": 4, "although": 4, "advis": 4, "choos": 4, "coli": 4, "b": 4, "subtili": 4, "gram": 4, "neg": 4, "bacteria": 4, "respect": 4, "str": 4, "folder": [4, 5], "is_refer": 4, "bool": 4, "coordin": 4, "other": 4, "object": [4, 5], "meprocess": 4, "metroubleshoot": 4, "attribut": 4, "manual": [4, 5], "curat": [4, 5], "file": [4, 5], "moreov": 4, "call": 4, "base": 4, "almost": 4, "origin": 4, "ecolim": [4, 5], "code": 4, "py": 4, "adapt": 4, "were": 4, "builder": 4, "final": 4, "version": 4, "predict": 4, "infer": 4, "reciproc": 4, "hit": 4, "blast": 4, "result": 4, "updat": 4, "its": 4, "evalu": 4, "float": 4, "cutoff": 4, "homolog": 4, "two": 4, "less": 4, "than": 4, "comprehens": 5, "pipelin": 5, "integr": 5, "packag": 5, "cobram": 5, "solvem": 5, "function": 5, "implement": 5, "prokaryot": 5, "readili": 5, "organ": 5, "input": 5, "work": 5, "four": 5, "synchron": 5, "contradictori": 5, "entri": 5, "homologi": 5, "templat": 5, "complet": 5, "user": 5, "step": 5, "resourc": 5, "intend": 5, "describ": 5, "basic": 5, "demonstr": 5, "get": 5, "start": 5, "type": 5, "mereconstruct": 5, "frequent": 5, "ask": 5, "question": 5, "my": 5, "what": 5, "modul": 5, "page": 5}, "objects": {}, "objtypes": {}, "objnames": {}, "titleterms": {"descript": [0, 5], "input": [0, 2, 3], "type": 0, "requir": 0, "option": 0, "configur": 0, "genom": 0, "gb": 0, "m": [0, 1], "model": [0, 1, 2, 3], "m_model": 0, "json": [0, 2], "gene": [0, 1, 2], "dictionari": 0, "txt": [0, 2], "note": 0, "protein": [0, 2], "rna": [0, 2], "tu": [0, 2], "sequenc": [0, 2], "fasta": [0, 2], "path": 0, "file": [0, 1, 2, 3], "paramet": [0, 2, 4], "organ": [0, 2, 4], "frequent": 1, "ask": 1, "question": 1, "my": 1, "identifi": 1, "ar": 1, "consist": 1, "what": 1, "should": 1, "i": 1, "do": 1, "1": 1, "genbank": 1, "thei": 1, "biocyc": [1, 2], "2": 1, "get": 2, "start": 2, "download": 2, "from": 2, "go": 2, "tool": 2, "special": 2, "smarttabl": 2, "initi": 2, "folder": 2, "your": 2, "defin": 2, "reconstruct": 2, "coralm": [2, 3, 4, 5], "import": 2, "packag": 2, "creat": 2, "builder": 2, "gener": 2, "build": 2, "me": [2, 3], "troubleshoot": 2, "curat": [2, 3], "manual": [2, 3], "how": 3, "us": 3, "arquitectur": 4, "mebuild": 4, "mereconstruct": 4, "homologi": 4, "welcom": 5, "document": 5, "content": 5, "indic": 5, "tabl": 5}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinx.ext.todo": 2, "sphinx.ext.viewcode": 1, "nbsphinx": 4, "sphinx": 57}, "alltitles": {"Description of Inputs": [[0, "Description-of-Inputs"]], "Types of inputs": [[0, "Types-of-inputs"]], "Required": [[0, "Required"]], "Optional": [[0, "Optional"]], "Configuration": [[0, "Configuration"]], "Description": [[0, "Description"], [5, "description"]], "Genome (genome.gb)": [[0, "Genome-(genome.gb)"]], "Requirements": [[0, "Requirements"], [0, "id1"], [0, "id2"], [0, "id3"], [0, "id5"], [0, "id7"], [0, "id9"], [0, "id11"], [0, "id13"]], "M-model (m_model.json)": [[0, "M-model-(m_model.json)"]], "Gene dictionary (genes.txt) [optional]": [[0, "Gene-dictionary-(genes.txt)-[optional]"]], "Notes": [[0, "Notes"], [0, "id4"], [0, "id6"], [0, "id8"], [0, "id10"], [0, "id12"], [0, "id14"]], "Proteins (proteins.txt) [optional]": [[0, "Proteins-(proteins.txt)-[optional]"]], "RNAs (RNAs.txt) [optional]": [[0, "RNAs-(RNAs.txt)-[optional]"]], "TUs (TUs.txt) [optional]": [[0, "TUs-(TUs.txt)-[optional]"]], "Gene sequences (sequences.fasta) [optional]": [[0, "Gene-sequences-(sequences.fasta)-[optional]"]], "Configuration of paths to files (inputs.json)": [[0, "Configuration-of-paths-to-files-(inputs.json)"]], "Configuration of parameters (organism.json)": [[0, "Configuration-of-parameters-(organism.json)"]], "Frequently Asked Questions": [[1, "Frequently-Asked-Questions"]], "My gene identifiers are not consistent, what should I do?": [[1, "My-gene-identifiers-are-not-consistent,-what-should-I-do?"]], "1. M-model and Genbank are consistent, but they are not consistent with the BioCyc files.": [[1, "1.-M-model-and-Genbank-are-consistent,-but-they-are-not-consistent-with-the-BioCyc-files."]], "2. M-model and Genbank are not consistent": [[1, "2.-M-model-and-Genbank-are-not-consistent"]], "Getting started": [[2, "Getting-started"]], "Download files from BioCyc": [[2, "Download-files-from-BioCyc"]], "Go to Tools>Special SmartTables": [[2, "Go-to-Tools>Special-SmartTables"]], "Download genes.txt and sequences.fasta": [[2, "Download-genes.txt-and-sequences.fasta"]], "Download proteins.txt, RNAs.txt and TUs.txt": [[2, "Download-proteins.txt,-RNAs.txt-and-TUs.txt"]], "Download proteins.txt": [[2, "Download-proteins.txt"]], "Download RNAs.txt": [[2, "Download-RNAs.txt"]], "Download TUs.txt": [[2, "Download-TUs.txt"]], "Initialize the folder for your organism": [[2, "Initialize-the-folder-for-your-organism"]], "Define inputs in input.json.": [[2, "Define-inputs-in-input.json."]], "Define parameters in organism.json.": [[2, "Define-parameters-in-organism.json."]], "Reconstruct with coralME": [[2, "Reconstruct-with-coralME"]], "Import packages": [[2, "Import-packages"]], "Define organism and inputs": [[2, "Define-organism-and-inputs"]], "Create builder": [[2, "Create-builder"]], "Generate files": [[2, "Generate-files"]], "Build ME-model": [[2, "Build-ME-model"]], "Troubleshoot ME-model": [[2, "Troubleshoot-ME-model"]], "Curate manually": [[2, "Curate-manually"]], "How to manually curate a ME-model using coralME": [[3, "How-to-manually-curate-a-ME-model-using-coralME"]], "Manual input files": [[3, "Manual-input-files"]], "How to curate?": [[3, "How-to-curate?"]], "Arquitecture of coralME": [[4, "Arquitecture-of-coralME"]], "Organism()": [[4, "Organism()"]], "Parameters": [[4, "Parameters"], [4, "id1"], [4, "id2"]], "MEBuilder()": [[4, "MEBuilder()"]], "MEReconstruction()": [[4, "MEReconstruction()"]], "Homology()": [[4, "Homology()"]], "Welcome to coralME documentation!": [[5, "welcome-to-coralme-documentation"]], "Content": [[5, "content"]], "Indices and tables": [[5, "indices-and-tables"]]}, "indexentries": {}})